rfcs/xxxx-apple-auth-provider.md

```diff
@@ -0,0 +1,169 @@
+
+The second idea is to provide a simple process to configure Apple provider like others.
+
+# Example
+
+We need to change a couple of file, because Apple SignIn is a mixed with oAuth2 and OpenID authentication process.
+
+### 1. Add Apple Grant main configuration
+
+File `strapi-plugin-users-permissions/config/functions/bootstrap.js`
+```javascript
+...
+apple: {
+      enabled: false,
+      icon: 'apple',
+      key: '',
+      secret: '', // First difficulty, the "secret" is a generated JWT token than need to be recreate usually
+      callback: `${strapi.config.server.url}/auth/apple/callback`,
+      scope: ['name', 'email'], // basic scope, but need to send a POST request
+      nonce: true,
+      state: true,
+      custom_params: {
+        response_type: 'code id_token',
+        response_mode: 'form_post' // Force Grant to perform a POST response
+      }
+    },
+...
+```
+
+### 2. Add a `POST` routes to received callback response from Apple (copy of GET)
+
+File `strapi-plugin-users-permissions/config/routes.json`
+
+```json
+{
+      "method": "POST",
+      "path": "/connect/*",
+      "handler": "Auth.connect",
+      "config": {
+        "policies": ["plugins::users-permissions.ratelimit"],
+        "prefix": "",
+        "description": "Connect a provider",
+        "tag": {
+          "plugin": "users-permissions",
+          "name": "User"
+        }
+      }
+    },
+```
+
+### Add Apple provider API call in `providers` services :
+
+
+File `strapi-plugin-users-permissions/services/providers.js`
+```javascript
+...
+const jwt = require('jsonwebtoken'); // require jwt decode
+...
+// /!\ DRAFT /!\ 
+case 'apple': {
+      const apple = purest({
+        provider: 'apple',
+        config: {
+          apple: {
+            'https://appleid.apple.com': {
+              __domain: {
+                auth: {
+                  auth: { bearer: '[0]' },
+                },
+              },
+              '{endpoint}': {
+                __path: {
+                  alias: '__default',
+                },
+              },
+            },
+          },
+        }
+      });
+      apple
+        .query()
+        .post('auth/token') // Apple want receive only post request
+        .auth(access_token)
+        .request((err, res, body) => {
+          if (err) {
+            callback(err);
+          } else {
+            const idToken = jwt.decode(body.id_token);
+            callback(null, {
+              username: idToken.email.split('@')[0],
+              email: idToken.email,
+            });
+          }
+        });
+      break;
+    }
+```
+
+This part is a draft.
+
+`Purest` has not been update since 4 years, many new providers are not available (like `discord` actually).
+It's a sample config for Apple than want receive a Bearer access token.
+
+Maybe it will be necessary to implement other library than `Purest`.
```

Comment:
Eventually we do plan to move the Users & Permissions plugin to passport.js and split all auth providers into some kind of `strapi-provider-auth-google` package (similar to upload and email). But right now there is no ETA unless @alexandrebodin has something to add.

I would imagine we would likely do this around the same time we develop the API key feature (which will be the default and Users & Permissions will be optional)